---
title: "Midterm_P3"
author: "Tianyu Xia"
date: "2024-10-21"
output: html_document
---

# BIOCB 4381/6381 Midterm

## Due Tuesday, November 5th, 11:59 PM (via Canvas)

##### THIS IS AN EXAM: You are supposed to do it independently. You can refer to any materials, but no collaboration is allowed. You can post questions on Ed Discussion only when you need further clarification. Please follow the instructions and write your answers in this Rmarkdown document.

##### [Problem 3(f) is only for 6381 section. Although it will not count as bonus points for 4381 section, feel free to try and give your answers.]{.underline}

##### [For submission, please make sure that you submit both the Rmarkdown file (midterm_P3.Rmd) and the Knited html file (midterm_P3.html) on Canvas.]{.underline}

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
# load DESeq2 library (this takes some time)
#if (!requireNamespace("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")
#BiocManager::install("DESeq2")

library(DESeq2)
```

## Problem 3 Differential Expression Analysis

Now we are going to use **DESeq2** to perform the differential expression analysis and identify the significantly differentially expressed genes between different groups.

*NOTE: for all parameters that are not specifically mentioned, just use the default ones*

The dataset we will use contains RNA-Seq data from 3 different tissue types (normal tissue, primary colorectal cancer and colorectal cancer metastatic in the liver) and 18 different individuals (so there are 54 samples in total). The data includes:

-   `count_data.csv`: read counts per gene in each sample

    -   row: gene
    -   column: sample

-   `sample_info.csv`: sample information including tissue types and individual IDs

### (a) Create a `DESeqDataSet` object

Please create a `DESeqDataSet` object named `dds`. For design formula, we want to test the effect of tissue type on expression level while accounting for individual IDs (no need to consider interaction term).

```{r} 
count_data <- read.csv("count_data.csv", row.names = 1)
sample_info <- read.csv("sample_info.csv", row.names = 1)
dds <- DESeqDataSetFromMatrix(countData = count_data,
                              colData = sample_info,
                              design = ~ tissueType + individualID)
```

```{r}

```

```{r}

```

### (b) Filter counts and run `DESeq2`

Before running `DESeq2`, we'd like to do some preliminary filtering of the data in dds: only include genes that have \>= 10 read counts across all samples in total (reads in sample 1 + reads in sample 2 + ... + reads in sample N \>= 10). Store the filtered data back to the same variable `dds` .

After the filtering, run `DESeq2` pipeline.

```{r}
dds <- dds[rowSums(counts(dds)) >= 10, ]
num_genes <- nrow(dds)
num_genes
dds <- DESeq(dds)

```

```{r}

```

```{r}

```

-   How many genes are left after filtering?

    Your answer:32710

### (c) Extract DESeq2 results

We'd like to compare gene expression level between **"primary_tumor"** group and **"normal"** group. Please identify **up-regulated genes** (log2FoldChange \> 2 and adjusted p-value \< 0.05) and **down-regulated genes** (log2FoldChange \< -2 and adjusted p-value \< 0.05) in "primary_tumor" group compared to "normal" group, and store them in two variables named `upregulated_genes` (data type: vector) and `downregulated_genes` (data type: vector).

Sort genes in each vector by adjusted p-values in an ascending order.

```{r}
result <- results(dds, contrast = c("tissueType", "primary_tumor", "normal"))
upregulated_genes <- subset(result, log2FoldChange > 2 & padj < 0.05)
upregulated_genes <- upregulated_genes[order(upregulated_genes$padj), ]
upregulated_genes <- rownames(upregulated_genes)
downregulated_genes <- subset(result, log2FoldChange < -2 & padj < 0.05)
downregulated_genes <- downregulated_genes[order(downregulated_genes$padj), ]
downregulated_genes <- rownames(downregulated_genes)
upregulated <- length(upregulated_genes)
downregulated <- length(downregulated_genes)
upregulated_genes
downregulated_genes
upregulated
downregulated
```

```{r}

```

```{r}

```

-   How many genes are significantly up-regulated and how many genes are significantly down-regulated?

    Your answer:371 and 516

### (d) Plot normalized counts for the most significant gene

Please find out the most significant gene with the smallest adjusted p-value. Please plot the normalized counts for this gene in 3 different tissue types.

```{r}
res <- result[order(result$padj), ]
most_sig_gene <- rownames(res)[1]
most_sig_gene
log2FoldChange <- res$log2FoldChange[1]
regulation <- ifelse(log2FoldChange > 0, "up-regulated", "down-regulated")
regulation
plotCounts(dds, gene = most_sig_gene, intgroup = "tissueType")
```

```{r}

```

-   What's the gene id of the most significant gene? Is it an up-regulated gene or a down-regulated gene?

    Your answer:ENSG00000142959 and it is down-regulated

### (e) Volcano plot

Finally, let's visualize the identified differentially expressed genes using a volcano plot. Please set those up-regulated genes as red dots and set those down-regulated genes as blue dots, and set other non-significant genes as black dots (using the cutoffs mentioned as before). Please also add the grey dashed lines as the cutoff lines (log2FoldChange = -2 or 2 and padj = 0.05) in the plot.

```{r}
mycols <- ifelse(result$log2FoldChange > 2 & result$padj < 0.05, "red",
                 ifelse(result$log2FoldChange < -2 & result$padj < 0.05, "blue", "black"))
plot(result$log2FoldChange, result$padj, col = mycols, pch = 20,
     xlab = "Log2FoldChange", ylab = "Adjusted p-value (padj)", main = "Volcano Plot")
abline(v = c(-2, 2), col = "gray", lty = 2)
abline(h = 0.05, col = "gray", lty = 2)
```

### (f) (BIOCB 6381 students only) EdgeR analysis

Learning how to use a new tool is important for computational biologists; thus, in this question you will use all the online resources to run the **EdgeR** analysis using the same data provided before.

Here we just focus on the different expression level between different tissue types and only want to test on **"primary_tumor"** group and **"normal"** group. You can ignore individual ID here.

*HINT: This [website](https://rnnh.github.io/bioinfo-notebook/docs/DE_analysis_edgeR_script.html) is good resources for you to learn hot to run EdgeR with the data that we created for `DESeq2`* .

Please follow the steps here:

1.  Create a DGEList object using `DGEList()` and add sample grouping information

2.  Filtering lowly expressed genes using `filterByExpr()`

3.  Normalizing samples using `calcNormFactors()`

4.  Estimating dispersion using `estimateDisp()`

5.  Run pairwise testing using `exactTest()`

6.  Use `topTags()` to get the most differentially expressed genes and display the results of the top 10 genes

```{r, include=FALSE}
# Load libraries (please install them if you haven't installed them)
#if (!requireNamespace("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")
BiocManager::install("limma")
BiocManager::install("edgeR")

library(limma)
library(edgeR)
```

```{r}
group <- factor(sample_info$tissueType)
dge <- DGEList(counts = count_data, group = group)
keep <- filterByExpr(dge, group = group)
dge <- dge[keep, , keep.lib.sizes = FALSE]
dge <- calcNormFactors(dge)
dge <- estimateDisp(dge)
test <- exactTest(dge, pair = c("normal", "primary_tumor"))
top_genes <- topTags(test, n = 10)
top_genes$table
```

# Congradulation! You have finished all the questions for the Midterm!
